src/snowflake/snowpark/_internal/data_source/drivers/psycopg2_driver.py

```diff
@@ -0,0 +1,362 @@
+                )
+            snow_type = BASE_POSTGRES_TYPE_TO_SNOW_TYPE.get(type_code)
+            if snow_type is None:
+                raise NotImplementedError(
+                    f"Postgres type not supported: {type_code} for column: {name}"
+                )
+            if Psycopg2TypeCode(type_code) == Psycopg2TypeCode.NUMERICOID:
+                if not self.validate_numeric_precision_scale(precision, scale):
+                    logger.debug(
+                        f"Snowpark does not support column"
+                        f" {name} of type {type_code} with precision {precision} and scale {scale}. "
+                        "The default Numeric precision and scale will be used."
+                    )
+                    precision, scale = None, None
+                data_type = snow_type(
+                    precision if precision is not None else 38,
+                    scale if scale is not None else 0,
+                )
+            elif type_code == Psycopg2TypeCode.TIMESTAMPTZOID:
+                data_type = snow_type(TimestampTimeZone.TZ)
+            else:
+                data_type = snow_type()
+            fields.append(StructField(name, data_type, True))
+        return StructType(fields)
+
+    @staticmethod
+    def data_source_data_to_pandas_df(
+        data: List[Any], schema: StructType
+    ) -> "pd.DataFrame":
+        df = BaseDriver.data_source_data_to_pandas_df(data, schema)
+
+        variant_type_indexes = [
+            i
+            for i, field in enumerate(schema.fields)
+            if isinstance(field.datatype, VariantType)
+        ]
+        col_names = df.columns[variant_type_indexes]
+        df[col_names] = BaseDriver.df_map_method(df[col_names])(
+            lambda x: json.dumps(x, cls=PythonObjJSONEncoder)
+        )
+
+        return df
+
+    @staticmethod
+    def to_result_snowpark_df(
+        session: "Session", table_name, schema, _emit_ast: bool = True
+    ) -> "DataFrame":
+        project_columns = []
+        for field in schema.fields:
+            if isinstance(field.datatype, VariantType):
+                project_columns.append(
+                    to_variant(parse_json(column(field.name))).as_(field.name)
+                )
+            else:
+                project_columns.append(column(field.name))
+        return session.table(table_name, _emit_ast=_emit_ast).select(
+            project_columns, _emit_ast=_emit_ast
+        )
+
+    @staticmethod
+    def prepare_connection(
+        conn: "Connection",
+        query_timeout: int = 0,
+    ) -> "Connection":
+        # The following is to align with Snowflake Connector behavior which get Interval as string
+        # the default behavior of psycopg2 is to get Interval as datetime.timedelta
+        # https://other-docs.snowflake.com/en/connectors/postgres6/view-data#postgresql-to-snowflake-data-type-mapping
+        from psycopg2.extensions import new_type, register_type
+
```

Comment:
I am thinking about how we should do this kind of type conversion and what is the best way. I can think of three place to do this:
1. in the select query, we may be able to do something like `to_char(interval_col) as interval_col `
2. in pandas dataframe, we could do type conversion with pandas map
3. if the driver have type converter, we could do conversion there like you did here

we should have a time discuss and test which is the best place to do conversions(or maybe a combination of it)

Reply:
switched to pushing down